src/StackExchange.Redis/Configuration/DefaultOptionsProvider.cs

```diff
@@ -0,0 +1,197 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Net;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace StackExchange.Redis.Configuration
+{
+    /// <summary>
+    /// A defaults providers for <see cref="ConfigurationOptions"/>.
+    /// This providers defaults not explicitly specified and is present to be inherited by environments that want to provide
+    /// better defaults for their use case, e.g. in a single wrapper library used many places.
+    /// </summary>
+    /// <remarks>
+    /// Why not just have a default <see cref="ConfigurationOptions"/> instance? Good question!
+    /// Since we null coalesce down to the defaults, there's an inherent pit-of-failure with that approach of <see cref="StackOverflowException"/>.
+    /// If you forget anything or if someone creates a provider nulling these out...kaboom.
+    /// </remarks>
+    public class DefaultOptionsProvider
+    {
+        /// <summary>
+        /// Whether this options provider matches a given endpoint, for automatically selecting a provider based on what's being connected to.
+        /// </summary>
+        public virtual bool IsMatch(EndPoint endpoint) => false;
+
+        /// <summary>
+        /// Gets a provider for the given endpoints, falling back to <see cref="DefaultOptionsProvider"/> if nothing more specific is found.
+        /// </summary>
+        internal static Func<EndPointCollection, DefaultOptionsProvider> GetForEndpoints { get; } = (endpoints) =>
+        {
+            foreach (var endpoint in endpoints)
+            {
+                foreach (var provider in KnownProviders)
+                {
+                    if (provider.IsMatch(endpoint))
+                    {
+                        return provider;
+                    }
+                }
+            }
+
+            return new DefaultOptionsProvider();
+        };
+
+        /// <summary>
+        /// Known providers to match - this is intentionally modifiable for expert users and wrapper libraries.
+        /// </summary>
+        public static List<DefaultOptionsProvider> KnownProviders { get; set; } = new()
```

Comment:
I think this needs some thought, but yes that's the intent: added ones take precedence. I'm thinking: last one wins - so maybe instead of a mutable list we have a .AddProvider() method, but I'd like to allow users to remove one...I think? I dunno, want thoughts here and do not think it's quite right as-is.

Reply:
I like the .AddProvider() route - we can probably live without the ability to remove providers. If people want different defaults, they'll be creating a fresh new connection anyway

Reply:
Alrighty, pushed this up for eyes! I'd like to add a dummy wrapper lib in the solution to prove all this out still - not feeling well today so maybe tomorrow evening.